op-program/l2/engineapi/l2_engine_api.go

```diff
@@ -116,17 +212,8 @@ func (ea *L2EngineAPI) endBlock() (*types.Block, error) {
 	ea.l2BuildingHeader = nil
 
 	header.GasUsed = header.GasLimit - uint64(*ea.l2GasPool)
-	header.Root = ea.l2BuildingState.IntermediateRoot(ea.l2Cfg.Config.IsEIP158(header.Number))
+	header.Root = ea.l2BuildingState.IntermediateRoot(ea.backend.Config().IsEIP158(header.Number))
 	block := types.NewBlock(header, ea.l2Transactions, nil, ea.l2Receipts, trie.NewStackTrie(nil))
-
-	// Write state changes to db
-	root, err := ea.l2BuildingState.Commit(ea.l2Cfg.Config.IsEIP158(header.Number))
-	if err != nil {
-		return nil, fmt.Errorf("l2 state write error: %w", err)
-	}
-	if err := ea.l2BuildingState.Database().TrieDB().Commit(root, false); err != nil {
-		return nil, fmt.Errorf("l2 trie write error: %w", err)
-	}
```

Comment:
Building a new block via `forkChoiceUpdated`/`getPayload` shouldn't actually make any changes to the database - that only happens when the block is actually imported with `NewPayload`.  This commit was previously required because the separate `ea.l2Database` was used when building blocks and that's not connected to the database that actually imports blocks, so if we didn't commit the changes here, we couldn't build another block on top of this one.

However, we now correctly use `ea.backend.StateAt` which uses a database that will include changes from imported blocks and hence don't need to commit when building.

Downside is that for the oracle backed version we now need to re-apply the block when importing it which previously we could have skipped (as long as all blocks were created locally) because we'd already committed the database changes when building it.